### PR TITLE
feat: [SPG-856]: add artifactStream_createdAt_buildNo index in artifact stream collection

### PR DESCRIPTION
**Pull request by inderschera**
## Describe your changes
Issue: Query: find { find: "artifacts", filter: { artifactStreamId: "lGs1XrKJSYyT7vScn1oIBQ", metadata.buildNo: /-master$/ }, sort: { createdAt: -1 }, limit: 1, planSummary: IXSCAN { createdAt: 1 } keysExamined:37111 docsExamined:37111 fromMultiPlanner:1 replanned:1 replanReason:"cached plan was less efficient than expected: expected trial execution to take 2705 works but it took at least 27050 works"

To solve this, added index artifacts {"name": "artifactStream_createdAt_buildNo", "background": true} {"artifactStreamId": 1, "createdAt": -1, "metadata.buildNo": 1} following the esr rule.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)

### CODE REVIEW

File: 957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java [hunk 1]
```diff
@@ -76,6 +76,12 @@ public static List<MongoIndex> mongoIndexes() {
                  .field(ArtifactKeys.artifactStreamId)
                  .field(ArtifactKeys.metadata_buildNo)
                  .build())
+        .add(SortCompoundMongoIndex.builder()
+                 .name("artifactStream_createdAt_buildNo")
+                 .field(ArtifactKeys.artifactStreamId)
+                 .descSortField(ArtifactKeys.createdAt)
+                 .field(ArtifactKeys.metadata_buildNo)
```
Comment (lasdox) [inline 957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java:83]: why not to use a text index here? @inderschera

**Comment by abhinav-harness** [comment]
*[`957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java` line 83 at r3](https://reviewable.io/reviews/harness/harness-core/40299#-NH4QecM6iz0o1AhxjfJ-r3-83:-NH5Wqle44V_q7EvkS1V:b83htva) ([raw file](https://github.com/harness/harness-core/blob/ba34823dc64ad41eb8d4a1ad46e5658009540089/957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java#L83)):*
<details><summary><i>Previously, lasdox (Lucas Sales) wrote…</i></summary><blockquote>

why not to use a text index here? @inderschera 
</blockquote></details>

@lasdox it can be overkill here, no?
also we can't do in between the text search with text index which is the use case with artifcat

**Comment by lasdox** [comment]
> _[`957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java` line 83 at r3](https://reviewable.io/reviews/harness/harness-core/40299#-NH4QecM6iz0o1AhxjfJ-r3-83:-NH5Wqle44V_q7EvkS1V:b83htva) ([raw file](https://github.com/harness/harness-core/blob/ba34823dc64ad41eb8d4a1ad46e5658009540089/957-cg-beans/src/main/java/software/wings/beans/artifact/Artifact.java#L83)):_
> 
> Previously, lasdox (Lucas Sales) wrote…
> @lasdox it can be overkill here, no? also we can't do in between the text search with text index which is the use case with artifcat

Then I see no point in creating another index here.
We have already `accountId_artifactStreamId_createdAt` we could just add accountId to the query and it would be fixed.
Also regex will not use index.

**Review by inderschera — COMMENTED** [review]
@lasdox regex do use an index. I was checking this: https://www.mongodb.com/docs/manual/reference/operator/query/regex/#index-use
But you have a good point. Will it make sense to add metadata_buildNo to last of the above `accountId_artifactStreamId_createdAt` index?
In any case I will add accountId filter in the query.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299#-:-NHNwgab6tbK3eq558c-:brlzgsi)* status: Some of the checks are too old. 2 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained.

**Comment by lasdox** [comment]
> @lasdox regex do use an index. I was checking this: https://www.mongodb.com/docs/manual/reference/operator/query/regex/#index-use But you have a good point. Will it make sense to add metadata_buildNo to last of the above `accountId_artifactStreamId_createdAt` index? In any case I will add accountId filter in the query.
> 
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299#-:-NHNwgab6tbK3eq558c-:brlzgsi)_ status: Some of the checks are too old. 2 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained.

@inderschera 
yes you are right, I agree, adding `metadata_buildNo` would help then.
we also have to be careful to not overkill it. 
I'll let you decide, I have no strong opinions on this.

**Review by hextechpal — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)* status: 1. unresolved discussions.

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)* status: 1. unresolved discussions.

**Review by Fernando-Dourado — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)* status: 1. unresolved discussions.

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)* status: 1. unresolved discussions.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40299)* status: 1. unresolved discussions.

**Pull request merged**